proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -0,0 +1,121 @@
+
+Lambdas or method groups with natural types can be used as initializers in `var` declarations.
+
+If the lambda or method group has no more than 16 parameters and no return value, and all parameters are passed by value, the natural type will be `delegate void System.Action<P1, ..., Pn>(P1, ..., Pn)` where `P1, ..., Pn` are the lambda parameter types.
+
+If the lambda or method group has no more than 16 parameters and a by-value return type, and all parameters are passed by value, the natural type will be `delegate R System.Func<P1, ..., Pn, R>(P1, ..., Pn)` where `P1, ..., Pn` are the lambda parameter types, and `R` is the lambda return type.
+
+Otherwise the natural type will be a synthesized `internal` anonymous `delegate` type with a signature that matches the lambda or method group.
+
+```csharp
+var f1 = () => default;        // error: no natural type
+var f2 = x => { };             // error: no natural type
+var f3 = x => x;               // error: no natural type
+var f4 = () => 1;              // System.Func<int>
+var f5 = () : string => null;  // System.Func<string>
+```
+
+```csharp
+static void F1() { }
+static void F1<T>(this T t) { }
+static void F2(this string s) { }
+
+var f6 = F1;    // error: multiple methods
+var f7 = "".F1; // System.Action
+var f8 = F2;    // System.Action<string> 
+```
+
+### Anonymous delegate type
+If synthesized delegate types are required, the compiler generates generic anonymous delegate types that are shared across all anonymous delegates in the module that have the same number of parameters and same parameter ref kinds.
+
+The names of the synthesized delegate types and the names of the parameters are unspeakable.
+
+The anonymous delegate types are not co- or contra-variant unlike the delegates constructed from `System.Action<>` and `System.Func<>`.
+
+The natural type of a method group does not include any `modopt()` or `modreq()` on the method group.
+
+`fA` and `fB` share a common generic type `internal delegate ref R <anonymous2><P1>(ref P1 <unnamed>)` below:
+```csharp
+var f9 = (ref int i) => { };    // delegate void <anonymous1><int>(ref int <unnamed>);
+var fA = (ref char c) => ref c; // delegate ref char <anonymous2><char, char>(ref char <unnamed>);
+var fB = (ref int i) => ref i;  // delegate ref int <anonymous2><int, int>(ref int <unnamed>);
+```
+
+_Issue: Is it necessary to use `System.Action<>` and `System.Func<>` for the delegate types or should the compiler always generate an anonymous delegate type? There is an observable difference since the anonymous delegate types are not co- or contra-variant._
+
+## Syntax
+
+```antlr
+lambda_expression
+  : attribute_list* modifier* lambda_parameters (':' type)? '=>' (block | body)
+  ;
+
+lambda_parameters
+  : lambda_parameter
+  | '(' (lambda_parameter (',' lambda_parameter)*)? ')'
+  ;
+
+lambda_parameter
+  : identifier
+  | (attribute_list* modifier* type)? identifier equals_value_clause?
```

Comment:
couold we do: `attribute_list? (modifier* type)? identifier equals_value_clause?` instead?

it seems like `([x] a) => ...` is reasonable, and doesn't require the user to have to provide full types for lambdas that want attributes.

Reply:
I'm of 2 minds on this:

1. We don't allow you specify just the `ref` today. If you want more signature than just a name, you have to have the whole signature. This decision is in line with that.
2. https://github.com/dotnet/csharplang/issues/338 is championed and would remove that restriction, so why add a similar restriction here?

Reply:
Exactly :)  but even if we don't do #338, i wouldn't perpetuate this.  :)

Reply:
Added an open issue for now.

---
In reply to: [579450645](https://github.com/dotnet/csharplang/pull/4451#discussion_r579450645) [](ancestors = 579450645)